### PR TITLE
HDDS-2534. scmcli container delete not working

### PR DESCRIPTION
**Pull request by chimney-lee**
## What changes were proposed in this pull request?

What changes were proposed in this pull request?
fix scmcli container delete not working

What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2534
#214 

How was this patch tested?
exec bin/ozone scmcli container delete  <containerId>

### CODE REVIEW

**Comment by chimney-lee** [comment]
@xiaoyuyao please review.  relates to #214

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java [hunk 1]
```diff
@@ -282,6 +290,12 @@ public ContainerInfo allocateContainer(final ReplicationType type,
   public void deleteContainer(ContainerID containerID) throws IOException {
     lock.lock();
     try {
+      Set<ContainerReplica> replicas =
+          containerStateManager.getContainerReplicas(containerID);
+      for(ContainerReplica replica : replicas){
+        nodeManager.removeContainers(replica.getDatanodeDetails(), Stream.of(
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java]: Can we create the set of containerID outside the loop with a simpler Java APIs:

Set<ContainerID> containerIdSet = Collections.singleton(containerID);
replias.forEach(replica->replica.removeContainers(replica.getDatanodeDetails(), containerIdSet));

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java [hunk 2]
```diff
@@ -293,6 +307,12 @@ public void deleteContainer(ContainerID containerID) throws IOException {
                 " it's missing!", containerID);
       }
       scmContainerManagerMetrics.incNumSuccessfulDeleteContainers();
+    } catch (NodeNotFoundException nnfe) {
+      scmContainerManagerMetrics.incNumFailureDeleteContainers();
+      throw new SCMException(
+          "Failed to delete container " + containerID + ", reason : " +
+              "node doesn't exist.",
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java]: Can we add the node info to the Exception? Is it part of the NodeNotFoundException?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerReportHandler.java [hunk 3]
```diff
@@ -119,21 +127,33 @@ public void onMessage(final ContainerReportFromDatanode reportFromDatanode,
...
+        containers.add(replicaProto);
       } catch (ContainerNotFoundException e) {
-        LOG.error("Received container report for an unknown container" +
-                " {} from datanode {}.", replicaProto.getContainerID(),
-            datanodeDetails, e);
+        LOG.info("Received container report for an unknown container" +
+                " {} from datanode {},delete it.",
+            replicaProto.getContainerID(),
+            datanodeDetails);
+        publisher.fireEvent(
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerReportHandler.java]: Should we fireEvent upon receive the scmcli request? Can we add some high level design description wrt the flow of the delete in the PR description? That will help the reviewers.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java [hunk 4]
```diff
@@ -563,6 +563,20 @@ public void setContainers(DatanodeDetails datanodeDetails,
...
+   *
+   * @param datanodeDetails - DatanodeID
+   * @param containerIds    - Set of containerIDs
+   * @throws NodeNotFoundException - if datanode is not known. For new datanode
+   *                               use addDatanodeInContainerMap call.
+   */
+  @Override
+  public void removeContainers(DatanodeDetails datanodeDetails,
+      Set<ContainerID> containerIds) throws NodeNotFoundException {
+    nodeStateManager.removeContainers(datanodeDetails.getUuid(), containerIds);
```
Comment (xiaoyuyao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java]: Do we need to remove the container from the pipelineContainerMap in SCMPipelineManager?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerReportHandler.java (hunk 3 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerReportHandler.java]: +1 for the description, the delete container flow is quite important for Recon to work correctly and hence trying to understand this better. Will there be a case where a user will need to delete a container directly using SCM?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java (hunk 4 above)
Comment (chimney-lee) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java]: Before delete a container, we should close it; When we close a container , The function SCMContainerManager#updateContainerState will remove the container from pipelineContainerMap in SCMPipelineManager.

**Comment by elek** [comment]
/pending Questions/suggestions from @xiaoyuyao are not yet addressed in #432 12f3f8ac94cf8808757bb7673e4208d8b0fede09

**Comment by maobaolong** [comment]
@chimney-lee Thank you for working on this, do you still working on?

**Comment by chimney-lee** [comment]
> @chimney-lee Thank you for working on this, do you still working on?

**Comment by chimney-lee** [comment]
@maobaolong   Sorry for this , it has been a long time, glad if someone could contine this work

**Comment by maobaolong** [comment]
@chimney-lee Thank you for the past working on this, i will continue this work.

**Comment by adoroszlai** [comment]
Thanks @chimney-lee for the work on this so far, and @maobaolong for taking it up.  I assume @maobaolong will open a new PR for the updated change, so closing this.

**Pull request closed without merging**